spectator/writer/udp_writer_test.go

```diff
@@ -0,0 +1,100 @@
+package writer
+
+import (
+	"github.com/Netflix/spectator-go/spectator/logger"
+	"net"
+	"testing"
+	"time"
+)
+
+func TestNewUdpWriter(t *testing.T) {
+	writer, err := NewUdpWriter("localhost:5000", logger.NewDefaultLogger())
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if writer == nil {
+		t.Errorf("Expected writer to be not nil")
+	}
+}
+
+func TestUdpWriter_Close(t *testing.T) {
+	writer, _ := NewUdpWriter("localhost:5000", logger.NewDefaultLogger())
+	err := writer.Close()
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+}
+
+func TestNewUdpWriter_InvalidAddress(t *testing.T) {
+	writer, err := NewUdpWriter("invalid address", logger.NewDefaultLogger())
+	if err == nil {
+		t.Errorf("Expected error, got nil")
+	}
+	if writer != nil {
+		t.Errorf("Expected writer to be nil")
+	}
+}
+
+// Test write after close using a local UDP server
+func TestUdpWriter_WriteAfterClose(t *testing.T) {
+	// Start a local UDP server
+	pc, err := net.ListenPacket("udp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Could not start UDP server: %v", err)
+	}
+	defer pc.Close()
+
+	// Create a new UDP writer
+	writer, err := NewUdpWriter(pc.LocalAddr().String(), logger.NewDefaultLogger())
+	if err != nil {
+		t.Fatalf("Could not create UDP writer: %v", err)
+	}
+
+	// Close the writer
+	_ = writer.Close()
+
+	// Write a message
+	writer.Write("test message")
+
+	// Check that no message was received
+	buffer := make([]byte, 1024)
+	_ = pc.SetReadDeadline(time.Now().Add(time.Second)) // prevent infinite blocking
+	_, _, err = pc.ReadFrom(buffer)
+	// ReadFrom will throw error if no message is received after timeout
+	if err == nil {
+		t.Errorf("Expected error, got nil")
+	}
+
+}
+
+func TestUdpWriter_Write(t *testing.T) {
```

Comment:
One thing I did with spectator-py was to write one unit test to see how many concurrent writes could be achieved with UDP.

https://github.com/Netflix/spectator-py/blob/main/tests/test_udpwriter.py

This test is disabled because it was unreliable on CI servers, but it helped provide an idea of what the library could achieve with sequential UDP writes. Not sure we need to have this here, but it could be useful, if we want to provide some guidance about how well this library will perform.

Reply:
Added a concurrent writes test.
Tested with 1,000,000 messages per thread with 4 threads and it completed successfully after 46 seconds.

Reply:
Talked too soon. If I take out the Debug log line on Write it starts failing with a few thousand messages per thread.
I guess the logger was giving time to the reader to catch up.